<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>QEco 2025 --- Allesina lecture notes – tutorial_glv_data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">QEco 2025 — Allesina lecture notes</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link active" href="../index.html" rel="" target="" aria-current="page">
 <span class="menu-text">Lecture notes</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../about.html" rel="" target="">
 <span class="menu-text">About</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools ms-auto">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../tutorial/tutorial_GLV_data.html">GLV and experimental data — Hands on tutorial</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">QEco 2025</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lecture_1.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">The Generalized Lotka-Volterra model</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lecture_2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">GLV with random parameters</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lecture_3.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">GLV and experimental data</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../tutorial/tutorial_GLV_data.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">GLV and experimental data — Hands on tutorial</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#glv-and-experimental-data-hands-on-tutorial" id="toc-glv-and-experimental-data-hands-on-tutorial" class="nav-link active" data-scroll-target="#glv-and-experimental-data-hands-on-tutorial">GLV and experimental data — Hands on tutorial</a>
  <ul class="collapse">
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data">Data</a></li>
  <li><a href="#loading-the-code" id="toc-loading-the-code" class="nav-link" data-scroll-target="#loading-the-code">Loading the code</a></li>
  <li><a href="#sum-of-squares" id="toc-sum-of-squares" class="nav-link" data-scroll-target="#sum-of-squares">Sum of squares</a></li>
  <li><a href="#weighted-least-squares" id="toc-weighted-least-squares" class="nav-link" data-scroll-target="#weighted-least-squares">Weighted least squares</a></li>
  <li><a href="#maximum-likelihood" id="toc-maximum-likelihood" class="nav-link" data-scroll-target="#maximum-likelihood">Maximum likelihood</a></li>
  <li><a href="#leave-one-out-cross-validation" id="toc-leave-one-out-cross-validation" class="nav-link" data-scroll-target="#leave-one-out-cross-validation">Leave-one-out cross validation</a></li>
  <li><a href="#simplified-models" id="toc-simplified-models" class="nav-link" data-scroll-target="#simplified-models">Simplified models</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">



<section id="glv-and-experimental-data-hands-on-tutorial" class="level1">
<h1>GLV and experimental data — Hands on tutorial</h1>
<section id="data" class="level2">
<h2 class="anchored" data-anchor-id="data">Data</h2>
<p>For our explorations, we are going to use recent data from Ishizawa and colleagues, which you can find here:</p>
<p>H. Ishizawa, Y. Tashiro, D. Inoue, M. Ike, H. Futamata. <a href="https://doi.org/10.1073/pnas.2312396121"><em>Learning beyond-pairwise interactions enables the bottom–up prediction of microbial community structure</em></a> PNAS 121 (7) e2312396121 (2024).</p>
<p>The Authors inoculated duckweed (<em>Lemna minor</em>) with synthetic bacterial communities formed by all possible combinations of seven strains. To this end, they cultured the infected duckweed in flasks for 10 days. At the end of the experiment, they plated the communities on agar plates containing antibiotics that would allow the growth only of a particular strain. In this way, they were able to measure the final density of each of the seven strains in each of the <span class="math inline">\(2^7 - 1 = 127\)</span> possible communities, and conducted each experiment in replicate. The full data set reports the outcome of 692 separate experiments!</p>
<p>More modestly, here we are going to focus on a smaller pool of three strains taken from the seven available. We therefore have <span class="math inline">\(7\)</span> possible communities, ranging from a single strain growing in isolation to the three strains growing together.</p>
<p>Here’s the full table:</p>
<table class="table">
<thead>
<tr class="header">
<th>DW067</th>
<th>DW102</th>
<th>DW145</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td></td>
<td></td>
<td>0.8</td>
</tr>
<tr class="even">
<td></td>
<td></td>
<td>0.812</td>
</tr>
<tr class="odd">
<td></td>
<td></td>
<td>0.858</td>
</tr>
<tr class="even">
<td></td>
<td></td>
<td>0.874</td>
</tr>
<tr class="odd">
<td></td>
<td></td>
<td>0.854</td>
</tr>
<tr class="even">
<td></td>
<td></td>
<td>0.951</td>
</tr>
<tr class="odd">
<td></td>
<td></td>
<td>0.823</td>
</tr>
<tr class="even">
<td></td>
<td></td>
<td>0.922</td>
</tr>
<tr class="odd">
<td></td>
<td></td>
<td>1</td>
</tr>
<tr class="even">
<td></td>
<td>0.529</td>
<td></td>
</tr>
<tr class="odd">
<td></td>
<td>0.356</td>
<td></td>
</tr>
<tr class="even">
<td></td>
<td>0.532</td>
<td></td>
</tr>
<tr class="odd">
<td></td>
<td>0.409</td>
<td></td>
</tr>
<tr class="even">
<td></td>
<td>0.416</td>
<td></td>
</tr>
<tr class="odd">
<td></td>
<td>0.548</td>
<td></td>
</tr>
<tr class="even">
<td></td>
<td>0.645</td>
<td></td>
</tr>
<tr class="odd">
<td></td>
<td>0.41</td>
<td></td>
</tr>
<tr class="even">
<td></td>
<td>0.367</td>
<td></td>
</tr>
<tr class="odd">
<td></td>
<td>0.663</td>
<td>0.604</td>
</tr>
<tr class="even">
<td></td>
<td>0.703</td>
<td>0.909</td>
</tr>
<tr class="odd">
<td></td>
<td>0.395</td>
<td>0.661</td>
</tr>
<tr class="even">
<td></td>
<td>0.54</td>
<td>0.487</td>
</tr>
<tr class="odd">
<td></td>
<td>0.919</td>
<td>0.981</td>
</tr>
<tr class="even">
<td></td>
<td>0.707</td>
<td>0.828</td>
</tr>
<tr class="odd">
<td>0.729</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>0.781</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>0.811</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>1</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>0.811</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>0.793</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>0.617</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>0.919</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>0.645</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>0.513</td>
<td></td>
<td>0.719</td>
</tr>
<tr class="odd">
<td>0.419</td>
<td></td>
<td>0.428</td>
</tr>
<tr class="even">
<td>0.432</td>
<td></td>
<td>0.546</td>
</tr>
<tr class="odd">
<td>0.197</td>
<td></td>
<td>0.598</td>
</tr>
<tr class="even">
<td>0.249</td>
<td></td>
<td>0.904</td>
</tr>
<tr class="odd">
<td>0.242</td>
<td></td>
<td>0.558</td>
</tr>
<tr class="even">
<td>0.541</td>
<td>0.563</td>
<td></td>
</tr>
<tr class="odd">
<td>0.568</td>
<td>0.755</td>
<td></td>
</tr>
<tr class="even">
<td>0.371</td>
<td>0.707</td>
<td></td>
</tr>
<tr class="odd">
<td>0.566</td>
<td>0.524</td>
<td></td>
</tr>
<tr class="even">
<td>0.839</td>
<td>0.433</td>
<td></td>
</tr>
<tr class="odd">
<td>0.796</td>
<td>0.609</td>
<td></td>
</tr>
<tr class="even">
<td>0.328</td>
<td>1</td>
<td>0.416</td>
</tr>
<tr class="odd">
<td>0.321</td>
<td>0.436</td>
<td>0.58</td>
</tr>
<tr class="even">
<td>0.378</td>
<td>0.57</td>
<td>0.381</td>
</tr>
<tr class="odd">
<td>0.386</td>
<td>0.602</td>
<td>0.563</td>
</tr>
<tr class="even">
<td>0.379</td>
<td>0.774</td>
<td>0.293</td>
</tr>
</tbody>
</table>
<p>We can therefore associate each measurement with a) the strain being measured, <span class="math inline">\(i\)</span>; b) the community in which <span class="math inline">\(i\)</span> was grown, <span class="math inline">\(k\)</span>; and c) the (biological) replicate experiment, <span class="math inline">\(r\)</span>.</p>
</section>
<section id="loading-the-code" class="level2">
<h2 class="anchored" data-anchor-id="loading-the-code">Loading the code</h2>
<p>We are going to use the code accompanying the paper</p>
<ul>
<li>Skwara, A., Lemos‐Costa, P., Miller, Z.R. and Allesina, S., 2023. <a href="https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.14028">Modelling ecological communities when composition is manipulated experimentally</a>. Methods in Ecology and Evolution, 14(2), pp.696-707.</li>
</ul>
<p>which you can find here</p>
<p><a href="https://github.com/StefanoAllesina/skwara_et_al_2022/"><code>https://github.com/StefanoAllesina/skwara_et_al_2022/</code></a></p>
<p>we have slightly massaged the code for this tutorial.</p>
<p>To load the code in <code>R</code> change to the directory where you have stored the tutorial, and source this code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">source</span>(<span class="st">"skwara_et_al_2022.R"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="sum-of-squares" class="level2">
<h2 class="anchored" data-anchor-id="sum-of-squares">Sum of squares</h2>
<p>First, we are going to use all the data to find a matrix <span class="math inline">\(B\)</span> that best encodes the density of the populations. To this end, we try to minimize the SSQ</p>
<p><span class="math display">\[
SSQ(B) = \sum_r \sum_k \sum_i \left(\tilde{x}^{(k,r)}_i - x^{(k)}_i \right)^2
\]</span></p>
<p>This can be accomplished by calling:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>full_SSQ <span class="ot">&lt;-</span> <span class="fu">run_model</span>(</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">datafile =</span> <span class="st">"Ishizawa_3_strains.csv"</span>, <span class="co"># csv file containing the data to be fit</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">"full"</span>, <span class="co"># estimate B allowing each coefficient to take the best value</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">goalf =</span> <span class="st">"SSQ"</span>, <span class="co"># minimize Sum of Squared Deviations</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="co"># start from Identity matrix</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">skipEM =</span> <span class="cn">TRUE</span>, <span class="co"># go directly to numerical optimization</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_results =</span> <span class="cn">TRUE</span> <span class="co"># plot the results</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`mutate_if()` ignored the following grouping variables:
• Column `comm`</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numerical search"
[1] 1.289597
[1] 1.289597
[1] 1.289597
[1] 1.289597
[1] 1.289597
[1] 1.289597
[1] 1.289597
[1] 1.289597
[1] 1.289597
[1] 1.289597</code></pre>
</div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The values being printed are the SSQ after each round of numerical optimization (in this case, the calculation converges immediately to the solution).</p>
<p>The boxplots show the data (points), as well as the corresponding boxplots, with the horizontal line being the median value of <span class="math inline">\(\tilde{x}_i^{(k,r)}\)</span> across <span class="math inline">\(r\)</span>, the triangle shows the location of the empirical mean. The circle marks the fitted mean value for the combination of strain/community, obtained computing <span class="math inline">\(\left(B^{(k,k)}\right)^{-1}1_{\|k\|}\)</span>. As you can see, we can find a matrix <span class="math inline">\(B\)</span> that recapitulates the observed means quite well.</p>
<p>The variable <code>full_SSQ</code> now contains:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(full_SSQ)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 8
 $ data_name    : chr "Ishizawa_3_strains"
 $ observed     : num [1:50, 1:3] 0 0 0 0 0 0 0 0 0 0 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : NULL
  .. ..$ : chr [1:3] "DW067" "DW102" "DW145"
 $ predicted    : num [1:50, 1:3] 0 0 0 0 0 0 0 0 0 0 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : NULL
  .. ..$ : chr [1:3] "DW067" "DW102" "DW145"
 $ variances    : num [1:50, 1:3] 0 0 0 0 0 0 0 0 0 0 ...
  ..- attr(*, "dimnames")=List of 2
  .. ..$ : NULL
  .. ..$ : chr [1:3] "DW067" "DW102" "DW145"
 $ B            : num [1:3, 1:3] 1.3 -0.479 0.875 0.246 2.143 ...
 $ pars         : num [1:9] 1.63166 0.02719 -1.28356 -0.00694 0.43581 ...
 $ goal_function: num 1.29
 $ goal_type    : chr "SSQ"</code></pre>
</div>
</div>
<p>Let’s plot the predicted vs.&nbsp;observed values:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pred_obs</span>(full_SSQ)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Where the points mark the predicted vs observed means, and the crosses the full data. The dashed line is the 1:1 line.</p>
</section>
<section id="weighted-least-squares" class="level2">
<h2 class="anchored" data-anchor-id="weighted-least-squares">Weighted least squares</h2>
<p>We can repeat the calculation, but this time trying to minimize</p>
<p><span class="math display">\[
WLS(B) = \sum_r \sum_k \sum_i \left(\frac{\tilde{x}^{(k,r)}_i - x^{(k)}_i}{\sigma^{(k)}_i} \right)^2
\]</span></p>
<p>by calling:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>full_WLS <span class="ot">&lt;-</span> <span class="fu">run_model</span>(</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">datafile =</span> <span class="st">"Ishizawa_3_strains.csv"</span>, <span class="co"># csv file containing the data to be fit</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">"full"</span>, <span class="co"># estimate B allowing each coefficient to take the best value</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">goalf =</span> <span class="st">"WLS"</span>, </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="co"># start from Identity matrix</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">skipEM =</span> <span class="cn">TRUE</span>, <span class="co"># go directly to numerical optimization</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_results =</span> <span class="cn">TRUE</span> <span class="co"># plot the results</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`mutate_if()` ignored the following grouping variables:
• Column `comm`</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numerical search"
[1] 67.07833
[1] 67.07833
[1] 67.07833
[1] 67.07833
[1] 67.07833
[1] 67.07833
[1] 67.07833
[1] 67.07833
[1] 67.07833
[1] 67.07833</code></pre>
</div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pred_obs</span>(full_WLS)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Notice that the points have moved slightly—this is because we are penalizing deviations differently depending on the measured variance (and thus points with a higher variance can depart more strongly from the 1:1 line).</p>
</section>
<section id="maximum-likelihood" class="level2">
<h2 class="anchored" data-anchor-id="maximum-likelihood">Maximum likelihood</h2>
<p>Now we take a different approach, and take the observations to be independent samples from a log-normal distribution:</p>
<p><span class="math display">\[
\tilde{x}_i^{(k,r)} \sim LN(x_i^{(k)}, \sigma_i)
\]</span></p>
<p>i.e., we take each variable to have a mean determined by the corresponding <span class="math inline">\(x_i^{(k)}\)</span>, and a variance parameter that depends only on strain identity.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>full_LN <span class="ot">&lt;-</span> <span class="fu">run_model</span>(</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">datafile =</span> <span class="st">"Ishizawa_3_strains.csv"</span>, <span class="co"># csv file containing the data to be fit</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">"full"</span>, <span class="co"># estimate B allowing each coefficient to take the best value</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">goalf =</span> <span class="st">"LikLN"</span>, </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="co"># start from Identity matrix</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">skipEM =</span> <span class="cn">TRUE</span>, <span class="co"># go directly to numerical optimization</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_results =</span> <span class="cn">TRUE</span> <span class="co"># plot the results</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`mutate_if()` ignored the following grouping variables:
• Column `comm`</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numerical search"
[1] -49.19108
[1] -49.19108
[1] -49.19108
[1] -49.19108
[1] -49.19108
[1] -49.19108
[1] -49.19108
[1] -49.19108
[1] -49.19108
[1] -49.19108</code></pre>
</div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pred_obs</span>(full_LN)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Also in this case, we obtain a good fit.</p>
</section>
<section id="leave-one-out-cross-validation" class="level2">
<h2 class="anchored" data-anchor-id="leave-one-out-cross-validation">Leave-one-out cross validation</h2>
<p>The ultimate test for this type of model is to be able to predict experimental results before running the experiment.</p>
<p>In our case, we can try to leave out one of the 7 communities, and see which model recover the best result:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>dt <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"Ishizawa_3_strains.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Rows: 50 Columns: 3
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (3): DW067, DW102, DW145

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># the last row contains an experiment with only DW067 and DW102</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>dt <span class="sc">%&gt;%</span> <span class="fu">slice</span>(<span class="dv">40</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 3
  DW067 DW102 DW145
  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 0.541 0.563     0</code></pre>
</div>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>LOO_LN <span class="ot">&lt;-</span> <span class="fu">run_model_LOO</span>(<span class="at">datafile =</span> <span class="st">"Ishizawa_3_strains.csv"</span>, <span class="at">model =</span> <span class="st">"full"</span>, <span class="at">goalf =</span> <span class="st">"LikLN"</span>, <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="at">skipEM =</span> <span class="cn">TRUE</span>, <span class="at">plot_results =</span> <span class="cn">TRUE</span>, </span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">LOO_row_num =</span>  <span class="dv">40</span> <span class="co"># exclude all experiments with this community</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`mutate_if()` ignored the following grouping variables:
• Column `comm`</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numerical search"
[1] -43.08928
[1] -43.08928
[1] -43.08928
[1] -43.08928
[1] -43.08928
[1] -43.08928
[1] -43.08928
[1] -43.08928
[1] -43.08928
[1] -43.08928</code></pre>
</div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>LOO_WLS <span class="ot">&lt;-</span> <span class="fu">run_model_LOO</span>(<span class="at">datafile =</span> <span class="st">"Ishizawa_3_strains.csv"</span>, <span class="at">model =</span> <span class="st">"full"</span>, <span class="at">goalf =</span> <span class="st">"WLS"</span>, <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="at">skipEM =</span> <span class="cn">TRUE</span>, <span class="at">plot_results =</span> <span class="cn">TRUE</span>, </span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">LOO_row_num =</span>  <span class="dv">40</span> <span class="co"># exclude all experiments with this community</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`mutate_if()` ignored the following grouping variables:
• Column `comm`</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numerical search"
[1] 56.05198
[1] 56.05198
[1] 56.05198
[1] 56.05198
[1] 56.05198
[1] 56.05198
[1] 56.05198
[1] 56.05198
[1] 56.05198
[1] 56.05198</code></pre>
</div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-12-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>LOO_SSQ <span class="ot">&lt;-</span> <span class="fu">run_model_LOO</span>(<span class="at">datafile =</span> <span class="st">"Ishizawa_3_strains.csv"</span>, <span class="at">model =</span> <span class="st">"full"</span>, <span class="at">goalf =</span> <span class="st">"SSQ"</span>, <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="at">skipEM =</span> <span class="cn">TRUE</span>, <span class="at">plot_results =</span> <span class="cn">TRUE</span>, </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>                        <span class="at">LOO_row_num =</span>  <span class="dv">40</span> <span class="co"># exclude all experiments with this community</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`mutate_if()` ignored the following grouping variables:
• Column `comm`</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numerical search"
[1] 1.042831
[1] 1.042831
[1] 1.042831
[1] 1.042831
[1] 1.042831
[1] 1.042831
[1] 1.042831
[1] 1.042831
[1] 1.042831
[1] 1.042831</code></pre>
</div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>All three models are quite successful at recovering the observed means for the community we left out.</p>
</section>
<section id="simplified-models" class="level2">
<h2 class="anchored" data-anchor-id="simplified-models">Simplified models</h2>
<p>We have seen in the lecture that this approach estimates all <span class="math inline">\(n^2\)</span> coefficients of the matrix <span class="math inline">\(B\)</span>. For the approach to be successful, we need to have observed enough combinations of populations growing together (each population must be present in at least <span class="math inline">\(n\)</span> experiments with distinct compositions, each pair of populations must appear in at least one final community). When the number of populations in the pool is large, and experiments are few, this approach is unfeasible. We can therefore try to simplify the model to reduce the number of parameters.</p>
<p>The idea brought forward by Skwara <em>et al.</em> is to approximate the matrix <span class="math inline">\(B\)</span> as the sum of a diagonal matrix and a low-rank matrix. For example, a version of the model with only <span class="math inline">\(n + 1\)</span> parameters reads:</p>
<p><span class="math display">\[
B = D(s) + \alpha 1_n1_n^T
\]</span></p>
<p>i.e., a model in which each diagonal element <span class="math inline">\(B_{ii} = s_i + \alpha\)</span> has its own parameter (and thus can take arbitrary values), while the off-diagonals are all the same (i.e., each population has the same “mean-field” effect on all others).</p>
<p>A more general model using <span class="math inline">\(3 n - 1\)</span> parameters reads:</p>
<p><span class="math display">\[
B = D(s) + u v^T
\]</span></p>
<p>in which the diagonal coefficients <span class="math inline">\(B_{ii} = s_i + u_i v_i\)</span> can take arbitrary values, while the off-diagonal elements <span class="math inline">\(B_{ij} = u_i v_j\)</span> are constrained; the effect of species <span class="math inline">\(j\)</span> on <span class="math inline">\(i\)</span> depends on two values: <span class="math inline">\(v_i\)</span> that measures how strongly species <span class="math inline">\(i\)</span> respond to the presence of other species, and <span class="math inline">\(v_j\)</span> that measures the magnitude of the typical effect of species <span class="math inline">\(j\)</span>. Either <span class="math inline">\(u\)</span> or <span class="math inline">\(v\)</span> can be taken as unitary (i.e., <span class="math inline">\(\sum_i v_i^2 = 1\)</span>) without loss of generality, thereby bringing the total number of parameters to <span class="math inline">\(3n-1\)</span>. When the number of populations is small, this approach does not lead to big gains (e.g., for <span class="math inline">\(n=3\)</span> we have 8 parameters instead of 9), but the reduction in parameters is substantial when <span class="math inline">\(n\)</span> becomes larger (e.g., for the whole data set, <span class="math inline">\(n=7\)</span>, and thus we would use 20 parameters instead of 49).</p>
<p>Another advantage of this approach is that we can write the inverse in linear time, thereby removing the only big computational hurdle of the approach:</p>
<p><span class="math display">\[\begin{aligned}
B^{-1} 1_n&amp;= (D(s) + u v^T)^{-1}1_n\\
&amp;= \left(D(s)^{-1} - \frac{1}{1 + v^TD(s)^{-1}u}D(s)^{-1} u v^TD(s)^{-1}\right)1_n\\ &amp;= D(s)^{-1}\left(1_n - \frac{v^T D(s)^{-1}1_n}{1 + v^TD(s)^{-1}u} u\right)
\end{aligned}
\]</span></p>
<p>These models can be derived from the consumer-resource (see the paper by Skwara <em>et al.</em>).</p>
<p>We can test different versions of the simplified model. A model with only the diagonal and an extra parameter performs very poorly:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>diag_a11 <span class="ot">&lt;-</span> <span class="fu">run_model</span>(</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">datafile =</span> <span class="st">"Ishizawa_3_strains.csv"</span>, <span class="co"># csv file containing the data to be fit</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">"diag_a11t"</span>,</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">goalf =</span> <span class="st">"LikLN"</span>, </span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="co"># start from Identity matrix</span></span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">skipEM =</span> <span class="cn">TRUE</span>, <span class="co"># go directly to numerical optimization</span></span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_results =</span> <span class="cn">TRUE</span> <span class="co"># plot the results</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`mutate_if()` ignored the following grouping variables:
• Column `comm`</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numerical search"
[1] -22.62175
[1] -22.62175
[1] -22.62175
[1] -22.62175
[1] -22.62175
[1] -22.62175
[1] -22.62175
[1] -22.62175
[1] -22.62175
[1] -22.62175</code></pre>
</div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pred_obs</span>(diag_a11)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>A model with symmetric interactions, <span class="math inline">\(B = D(s) + vv^T\)</span> does slightly better:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>diag_vvt <span class="ot">&lt;-</span> <span class="fu">run_model</span>(</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">datafile =</span> <span class="st">"Ishizawa_3_strains.csv"</span>, <span class="co"># csv file containing the data to be fit</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">"diag_vvt"</span>,</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">goalf =</span> <span class="st">"LikLN"</span>, </span>
<span id="cb33-5"><a href="#cb33-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="co"># start from Identity matrix</span></span>
<span id="cb33-6"><a href="#cb33-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">skipEM =</span> <span class="cn">TRUE</span>, <span class="co"># go directly to numerical optimization</span></span>
<span id="cb33-7"><a href="#cb33-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_results =</span> <span class="cn">TRUE</span> <span class="co"># plot the results</span></span>
<span id="cb33-8"><a href="#cb33-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`mutate_if()` ignored the following grouping variables:
• Column `comm`</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numerical search"
[1] -38.12933
[1] -38.36503
[1] -38.37226
[1] -38.38069
[1] -38.38508
[1] -38.39331
[1] -38.39618
[1] -38.39655
[1] -38.39655
[1] -38.39655</code></pre>
</div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pred_obs</span>(diag_vvt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>And, finally, the model with 8 parameters, <span class="math inline">\(B = D(s) + vw^T\)</span>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>diag_vwt <span class="ot">&lt;-</span> <span class="fu">run_model</span>(</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">datafile =</span> <span class="st">"Ishizawa_3_strains.csv"</span>, <span class="co"># csv file containing the data to be fit</span></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">"diag_vwt"</span>,</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">goalf =</span> <span class="st">"LikLN"</span>, </span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">pars =</span> <span class="cn">NULL</span>, <span class="co"># start from Identity matrix</span></span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">skipEM =</span> <span class="cn">TRUE</span>, <span class="co"># go directly to numerical optimization</span></span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">plot_results =</span> <span class="cn">TRUE</span> <span class="co"># plot the results</span></span>
<span id="cb37-8"><a href="#cb37-8" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`mutate_if()` ignored the following grouping variables:
• Column `comm`</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "numerical search"
[1] -34.87107
[1] -49.06619
[1] -49.06753
[1] -49.0681
[1] -49.0681
[1] -49.0681
[1] -49.0681
[1] -49.06811
[1] -49.06811
[1] -49.06811</code></pre>
</div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_pred_obs</span>(diag_vwt)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="tutorial_GLV_data_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>